foundations/src/panic/hook.rs

```diff
@@ -0,0 +1,72 @@
+//! Panic hook implementation for tracking panics.
+
+use std::panic::{self, PanicHookInfo};
+use std::sync::OnceLock;
+
+pub(crate) static HOOK_INSTALLED: OnceLock<()> = OnceLock::new();
+
+/// Install the panic hook.
+///
+/// Returns `true` if this is the first installation, `false` if the hook
+/// was already installed (subsequent calls are no-ops).
+///
+/// See the module-level docs for more information: [`crate::panic`]
+pub fn install_hook() -> bool {
+    let first_install = HOOK_INSTALLED.set(()).is_ok();
```

Comment:
same here with the AtomicBool. Or if you 100% want to make sure this is not called anywhere in parallel, use `get_or_init` and install the hook inside the init function.